### PR TITLE
Fix PHPStan level 2 problems

### PR DESCRIPTION
**Pull request by szepeviktor**
An improved and explicit version of #1022

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -31,7 +31,9 @@
         "php": "^7.2 || ^8.0"
     },
     "require-dev": {
-        "phpunit/phpunit": "^8.0 || ~9.4.4"
+        "phpunit/phpunit": "^8.0 || ~9.4.4",
+        "friendsofphp/php-cs-fixer": "^3.13",
+        "phpstan/phpstan": "^1.9"
```
Comment (szepeviktor) [inline composer.json:36]: I'll be okay.
Complete CI comes in the next PR.

File: src/Cluster/SlotMap.php [hunk 2]
```diff
@@ -144,14 +144,16 @@ public function offsetExists($slot)
...
     #[ReturnTypeWillChange]
     public function offsetGet($slot)
     {
-        if (isset($this->slots[$slot])) {
-            return $this->slots[$slot];
+        if (! isset($this->slots[$slot])) {
+            return null;
         }
+
+        return $this->slots[$slot];
```
Comment (szepeviktor) [inline src/Cluster/SlotMap.php:156]: Making happy path the expected one. https://github.com/szepeviktor/very-clean-code/blob/master/README.md#-happy-and-unhappy-path

File: src/Client.php [hunk 3]
```diff
@@ -42,7 +42,7 @@
  *
  * @template-implements \IteratorAggregate<string, static>
  */
-class Client implements ClientInterface, IteratorAggregate
+final class Client implements ClientInterface, IteratorAggregate
```
Comment (tillkruss) [inline src/Client.php:45]: This would be a breaking change, we can't do that.

Comment (szepeviktor) [reply src/Client.php:45]: Oh nooooooooooooooooooooooo!

There is a `new static()` in it which will wrong if someone extends `Client`.
Reverting .... 🏃🏻‍♂️

**Comment by szepeviktor** [comment]
Null coalescing operator was introduced in PHP 7.0.
```php
     * @return NodeConnectionInterface|null
     */
    public function getConnectionByAlias($alias)
    {
        return $this->aliases[$alias] ?? null;
    }
```
Now using it in predis!

**Comment by szepeviktor** [comment]
@tillkruss Can we add variadic params to method using `func_get_args()`?
```diff
- public function transaction(/* arguments */)
+ public function transaction(...$arguments)
```

**Comment by tillkruss** [comment]
> @tillkruss Can we add variadic params to method using `func_get_args()` in the next PR?

Sure.

File: composer.json (hunk 1 above)
Comment (tillkruss) [reply composer.json:36]: No need for these changes, please remove.

File: src/Connection/Parameters.php [hunk 4]
```diff
@@ -19,7 +19,7 @@
  *
  * {@inheritdoc}
  */
-class Parameters implements ParametersInterface
+final class Parameters implements ParametersInterface
```
Comment (tillkruss) [inline src/Connection/Parameters.php:22]: Breaking change as well.

**Comment by szepeviktor** [comment]
@tillkruss Ready to review.

File: src/Cluster/ClusterStrategy.php [hunk 5]
```diff
@@ -351,10 +361,8 @@ protected function getKeyFromGeoradiusCommands(CommandInterface $command)
                 }
             }
 
-            if ($this->checkSameSlotForKeys($keys)) {
-                return $arguments[0];
-            } else {
-                return;
+            if (!$this->checkSameSlotForKeys($keys)) {
+                return null;
```
Comment (szepeviktor) [inline src/Cluster/ClusterStrategy.php:365]: `return $arguments[0];` is here below 👇🏻

File: src/Command/RawCommand.php [hunk 6]
```diff
@@ -43,11 +43,11 @@ public function __construct($commandID, array $arguments = [])
      * Creates a new raw command using a variadic method.
      *
      * @param string $commandID Redis command ID
-     * @param string ...        Arguments list for the command
+     * param  string ...$args   Arguments list for the command
      *
      * @return CommandInterface
      */
-    public static function create($commandID /* [ $arg, ... */)
+    public static function create($commandID /* , $arg, ... */)
```
Comment (szepeviktor) [inline src/Command/RawCommand.php:50]: These non-existent parameters will be fixed in the next PR.

File: src/Client.php [hunk 7]
```diff
@@ -467,7 +467,7 @@ protected function createTransaction(array $options = null, $callable = null)
      * Creates a new publish/subscribe context and returns it, or starts its loop
      * inside the optionally provided callable object.
      *
-     * @param mixed ... Array of options, a callable for execution, or both.
+     * param mixed ...$arguments Array of options, a callable for execution, or both.
```
Comment (tillkruss) [inline src/Client.php:470]: This is wrong, missing `@`. Several times in other files too.

**Review by tillkruss — CHANGES_REQUESTED** [review]

File: src/Client.php (hunk 7 above)
Comment (szepeviktor) [reply src/Client.php:470]: Yes, I've removed it from non-existent parameters.

in the next PR they will be dealt with: https://github.com/predis/predis/pull/1025#issuecomment-1396075402

Comment (szepeviktor) [reply src/Client.php:470]: Should I include it in this PR?

Comment (tillkruss) [reply src/Client.php:470]: Yes, I don't want to merge broken phpdocs.

Comment (szepeviktor) [reply src/Client.php:470]: Continuing with 🍊🍊🍊 power ...

File: src/Command/RawCommand.php (hunk 6 above)
Comment (szepeviktor) [reply src/Command/RawCommand.php]: Fixed.

**Comment by szepeviktor** [comment]
@tillkruss Please review and merge this one.
I'd like to continue with a complete CI (called "Lint").

**Comment by tillkruss** [comment]
> I'd like to continue with a complete CI (called "Lint").

Yes, please. 

> note to self: compile phpiredis in CI https://github.com/shivammathur/setup-php/wiki/Add-extension-from-source

You can skip that, will depreciate phpiredis and replace it with [Relay](https://relay.so).

**Review by tillkruss — APPROVED** [review]

**Pull request merged**